### PR TITLE
Work on plumbing for contexts-based performance tests

### PR DESCRIPTION
**Pull request by agrippa**
This pull request adds one initial contexts-based performance test for uni-directional bandwidth. However, most of the diff is adding plumbing to the testing infrastructure to allow multi-threaded tests to be configured from the command line and set up some shared infrastructure for them.

### CODE REVIEW

File: test/performance/shmem_perf_suite/shmem_bw_nb_put_perf.c [hunk 1]
```diff
@@ -35,15 +35,19 @@
 
 #include <bw_common.h>
 
+#include <pthread.h>
```
Comment (kseager) [inline test/performance/shmem_perf_suite/shmem_bw_nb_put_perf.c]: do these includes need to be in this file? I'm not seeing where this dependency came from. I'd rather keep the thread specific elements separated out the way you have with OpenMP.

File: test/performance/shmem_perf_suite/uni_dir_ctx.h [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            int i;
+            const int thread_id = omp_get_thread_num();
+            const shmemx_ctx_t ctx = ctxs[thread_id];
+            char *src_buffer = srcs[thread_id];
+            char *dest_buffer = dests[thread_id];
+
+            for (i = 0; i < metric_info->trials + metric_info->warmup; i++) {
+                if (i == metric_info->warmup) {
+                    shmemx_ctx_quiet(ctx);
```
Comment (kseager) [inline test/performance/shmem_perf_suite/uni_dir_ctx.h]: why only a quiet for warm-up?

File: test/performance/shmem_perf_suite/Makefile.am [hunk 3]
```diff
@@ -46,15 +46,16 @@ check_PROGRAMS += \
...
+	shmem_ctx_bw_nb_put_perf
 endif
 endif
 
 NPROCS ?= 2
 LOG_COMPILER = $(TEST_RUNNER)
 
-AM_CPPFLAGS = -I$(top_builddir)/mpp
-AM_LDFLAGS = $(LIBTOOL_WRAPPER_LDFLAGS)
+AM_CPPFLAGS = -I$(top_builddir)/mpp -fopenmp
```
Comment (kseager) [inline test/performance/shmem_perf_suite/Makefile.am]: if ENABLE_OPENMP here too? for a += add?

**Comment by kseager** [comment]
I was mostly focused on the testing infrastructure review. Thanks for the additions, like the added input error handling. I agree that "shmem_init" was better suited to the common file.  Overall this looks good

File: test/performance/shmem_perf_suite/shmem_bw_nb_put_perf.c (hunk 1 above)
Comment (agrippa) [reply test/performance/shmem_perf_suite/shmem_bw_nb_put_perf.c]: Sorry about that, some leftover code from testing. Removed.

File: test/performance/shmem_perf_suite/uni_dir_ctx.h (hunk 2 above)
Comment (agrippa) [reply test/performance/shmem_perf_suite/uni_dir_ctx.h]: Thanks for pointing that out, not sure why I originally thought that was necessary. This shmemx_ctx_quiet would essentially be a no-op everytime we hit it anyway. Removing.

File: test/performance/shmem_perf_suite/Makefile.am (hunk 3 above)
Comment (agrippa) [reply test/performance/shmem_perf_suite/Makefile.am]: Fixed.

**Comment by kseager** [comment]
👍 from my end. Good job on this addition. @jdinan do you have anything to add?

**Comment by jdinan** [comment]
I'll merge master to contexts this afternoon.  That should get Travis working again so we can test this PR properly.

**Comment by jdinan** [comment]
Merged master branch updates with contexts and fixed merge conflicts on this PR.  Travis tests running...

**Comment by jdinan** [comment]
These changes have been integrated into SOS.

**Pull request closed without merging**